File: grain-research-dashboard/dashboard/compress_data.py
#!/usr/bin/env python3
"""
压缩 spreads.html 中的 EMBEDDED_DATA 数据
将年份数组中的 null 值移除，只保留有效数据点和其索引
"""

import re
import json
import sys

def compress_seasonal_data(data):
    """压缩季节性数据，移除 null 值"""
    if not isinstance(data, dict):
        return data
    
    result = {}
    
    # 检查是否是季节性格式的数据（有 labels 和 datasets）
    if 'labels' in data and 'datasets' in data:
        labels = data['labels']
        datasets = data['datasets']
        
        result['labels'] = labels
        result['datasets'] = {}
        
        for spread_name, year_data in datasets.items():
            result['datasets'][spread_name] = {}
            
            for year, values in year_data.items():
                if isinstance(values, list):
                    # 提取非null值及其索引
                    compressed = [[i, v] for i, v in enumerate(values) if v is not None]
                    result['datasets'][spread_name][year] = compressed
                else:
                    result['datasets'][spread_name][year] = values
        
        return result
    
    # 递归处理其他字段
    for key, value in data.items():
        if isinstance(value, dict):
            result[key] = compress_seasonal_data(value)
        else:
            result[key] = value
    
    return result

def extract_and_compress_html(input_file, output_file):
    """提取HTML中的EMBEDDED_DATA，压缩后放回"""
    
    # 读取文件
    with open(input_file, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # 找到 EMBEDDED_DATA 的开始和结束
    # 格式: const EMBEDDED_DATA = { ... };
    pattern = r'(const EMBEDDED_DATA = )(\{[\s\S]*?\n\});'
    match = re.search(pattern, content)
    
    if not match:
        print("错误: 未找到 EMBEDDED_DATA")
        return False
    
    prefix = match.group(1)  # "const EMBEDDED_DATA = "
    json_str = match.group(2)  # JSON数据
    
    print(f"原始JSON长度: {len(json_str)} 字符")
    
    # 解析JSON
    try:
        data = json.loads(json_str)
    except json.JSONDecodeError as e:
        print(f"JSON解析错误: {e}")
        return False
    
    # 压缩数据
    print("正在压缩数据...")
    compressed_data = compress_seasonal_data(data)
    
    # 重新序列化为JSON
    compressed_json = json.dumps(compressed_data, ensure_ascii=False, separators=(',', ':'))
    print(f"压缩后JSON长度: {len(compressed_json)} 字符")
    print(f"压缩率: {100 - 100*len(compressed_json)/len(json_str):.1f}%")
    
    # 替换内容
    new_content = content[:match.start()] + prefix + compressed_json + content[match.end():]
    
    # 写入输出文件
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(new_content)
    
    print(f"已保存到: {output_file}")
    return True

if __name__ == '__main__':
    # 使用绝对路径
    input_file = r'C:\Users\administer\WorkBuddy\20260315144604\grain-research-dashboard\dashboard\spreads.html'
    output_file = r'C:\Users\administer\WorkBuddy\20260315144604\grain-research-dashboard\dashboard\spreads_compressed.html'
    
    if len(sys.argv) > 1:
        input_file = sys.argv[1]
    if len(sys.argv) > 2:
        output_file = sys.argv[2]
    
    success = extract_and_compress_html(input_file, output_file)
    sys.exit(0 if success else 1)

File: grain-research-dashboard/dashboard/compress_data.js
const fs = require('fs');
const path = require('path');

function compressSeasonalData(data) {
    if (typeof data !== 'object' || data === null) {
        return data;
    }
    
    // 检查是否是季节性格式的数据（有 labels 和 datasets）
    if (data.labels && data.datasets) {
        const labels = data.labels;
        const datasets = data.datasets;
        
        const result = {
            labels: labels,
            datasets: {}
        };
        
        for (const spreadName in datasets) {
            const yearData = datasets[spreadName];
            result.datasets[spreadName] = {};
            
            for (const year in yearData) {
                const values = yearData[year];
                if (Array.isArray(values)) {
                    // 提取非null值及其索引
                    const compressed = values.map((v, i) => [i, v]).filter(pair => pair[1] !== null);
                    result.datasets[spreadName][year] = compressed;
                } else {
                    result.datasets[spreadName][year] = values;
                }
            }
        }
        
        return result;
    }
    
    // 递归处理其他字段
    const result = {};
    for (const key in data) {
        result[key] = compressSeasonalData(data[key]);
    }
    return result;
}

function extractAndCompressHtml(inputPath, outputPath) {
    // 读取文件
    let content = fs.readFileSync(inputPath, 'utf-8');
    
    // 找到 EMBEDDED_DATA 的开始和结束
    const pattern = /(const EMBEDDED_DATA = )(\{[\s\S]*?\n\});/;
    const match = content.match(pattern);
    
    if (!match) {
        console.error('错误: 未找到 EMBEDDED_DATA');
        process.exit(1);
    }
    
    const prefix = match[1];  // "const EMBEDDED_DATA = "
    const jsonStr = match[2];  // JSON数据
    
    console.log(`原始JSON长度: ${jsonStr.length} 字符`);
    
    // 解析JSON
    let data;
    try {
        data = JSON.parse(jsonStr);
    } catch (e) {
        console.error(`JSON解析错误: ${e.message}`);
        process.exit(1);
    }
    
    // 压缩数据
    console.log('正在压缩数据...');
    const compressedData = compressSeasonalData(data);
    
    // 重新序列化为JSON
    const compressedJson = JSON.stringify(compressedData);
    console.log(`压缩后JSON长度: ${compressedJson.length} 字符`);
    console.log(`压缩率: ${100 - 100*compressedJson.length/jsonStr.length:.1f}%`);
    
    // 替换内容
    content = content.replace(pattern, prefix + compressedJson + ';');
    
    // 写入输出文件
    fs.writeFileSync(outputPath, content, 'utf-8');
    
    console.log(`已保存到: ${outputPath}`);
}

// 主程序
const inputFile = 'C:\\Users\\administer\\WorkBuddy\\20260315144604\\grain-research-dashboard\\dashboard\\spreads.html';
const outputFile = 'C:\\Users\\administer\\WorkBuddy\\20260315144604\\grain-research-dashboard\\dashboard\\spreads_compressed.html';
const logFile = 'C:\\Users\\administer\\WorkBuddy\\20260315144604\\grain-research-dashboard\\dashboard\\compress_log.txt';

const fs = require('fs');

// 重定向console输出到文件
const logStream = fs.createWriteStream(logFile, {flags: 'a'});
console.log = function(...args) {
    logStream.write(args.join(' ') + '\n');
};
console.error = function(...args) {
    logStream.write('ERROR: ' + args.join(' ') + '\n');
};

console.log('开始压缩...');
try {
    extractAndCompressHtml(inputFile, outputFile);
    console.log('完成!');
} catch(e) {
    console.error('错误:', e.message);
    console.error(e.stack);
}
logStream.end();
